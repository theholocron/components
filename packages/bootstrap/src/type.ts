import * as React from "react";

export interface WithChildren {
	children?: React.ReactNode;
}
