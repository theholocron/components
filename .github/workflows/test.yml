name: Test

on:
  push

jobs:
  types:
    name: Check build and types
    runs-on: ubuntu-latest
    strategy:
      matrix:
        package:
          - packages/bootstrap
    steps:
      - uses: actions/checkout@v4
        name: Checkout repository
        with:
          fetch-depth: 0

      - uses: actions/setup-node@v4
        name: Setup Node.js @v20
        with:
          cache: npm
          node-version: 20.x

      - run: npm ci
        name: Install dependencies
        env:
          HUSKY: 0

      - run: npm rebuild typescript && npm run build:lib && npm run types -- --project tsconfig.build.json
        name: Build the library and types
        working-directory: ${{ matrix.package }}

  coverage:
    name: Run tests and collect coverage
    runs-on: ubuntu-latest
    strategy:
      matrix:
        package:
          - packages/bootstrap
    steps:
      - uses: actions/checkout@v4
        name: Checkout repository
        with:
          fetch-depth: 0

      - uses: actions/setup-node@v4
        name: Setup Node.js @v20
        with:
          cache: npm
          node-version: 20.x

      - run: npm ci
        name: Install dependencies
        env:
          HUSKY: 0

      - run: npx playwright install --with-deps
        name: Install Playwright

      - run: npm run build:storybook -- --quiet --output-dir=docs
        name: Build Storybook
        working-directory: ${{ matrix.package }}

      - run: |
          npx concurrently -k -s first -n "SB,TEST" -c "magenta,blue" \
          "npx http-server docs --port 6006 --silent" \
          "npx wait-on tcp:6006 && npm run test:storybook -- --coverage --junit --ci"
        name: Run tests with coverage
        working-directory: ${{ matrix.package }}

      - uses: codecov/codecov-action@v4
        name: Upload results to Codecov
        with:
          files: ./${{ matrix.package }}/coverage/coverage-final.json
          flags: ${{ matrix.package }}
          token: ${{ secrets.CODECOV_TOKEN }}

      - uses: codecov/test-results-action@v1
        name: Upload test results to Codecov
        if: ${{ !cancelled() }}
        with:
          working-directory: ${{ matrix.package }}
          token: ${{ secrets.CODECOV_TOKEN }}

  # Run visual and composition tests with Chromatic
  visual-and-composition:
    name: Test Visual and Composition
    runs-on: ubuntu-latest
    strategy:
      matrix:
        package:
          - packages/bootstrap
    steps:
      - uses: actions/checkout@v4
        name: Checkout repository
        with:
          fetch-depth: 0

      - uses: actions/setup-node@v4
        name: Setup Node.js @v20
        with:
          cache: npm
          node-version: 20.x

      - run: npm ci
        name: Install dependencies
        env:
          HUSKY: 0

      - uses: chromaui/action@v1
        name: Publish to Chromatic
        with:
          buildScriptName: build:storybook
          projectToken: ${{ secrets.CHROMATIC_PROJECT_TOKEN }}
          token: ${{ secrets.GITHUB_TOKEN }}
          workingDir: ${{ matrix.package }}
        env:
          NEXT_PUBLIC_ENVIRONMENT: "development"
          STORYBOOK_NEXT_PUBLIC_ENVIRONMENT: "development"

  # Run interaction and accessibility tests
  interaction-and-accessibility:
    name: Test Interactions and Accessibility
    runs-on: ubuntu-latest
    strategy:
      matrix:
        package:
          - packages/bootstrap
    steps:
      - uses: actions/checkout@v4
        name: Checkout repository
        with:
          fetch-depth: 0

      - uses: actions/setup-node@v4
        name: Setup Node.js @v20
        with:
          cache: npm
          node-version: 20.x

      - run: npm ci
        name: Install dependencies
        env:
          HUSKY: 0

      - run: npx playwright install --with-deps
        name: Install Playwright

      - run: npm run build:storybook -- --quiet --output-dir=docs
        name: Build Storybook
        working-directory: ${{ matrix.package }}

      - run: |
          npx concurrently -k -s first -n "SB,TEST" -c "magenta,blue" \
          "npx http-server docs --port 6006 --silent" \
          "npx wait-on tcp:6006 && npm run test:storybook -- --ci"
        name: Serve Storybook and run tests
        working-directory: ${{ matrix.package }}

  # Run user flow tests with Cypress
  user-flow:
    name: Test User Flow
    runs-on: ubuntu-latest
    strategy:
      fail-fast: false # https://github.com/cypress-io/github-action/issues/48
      matrix:
        containers: [1, 2] # Uses 2 parallel instances
        package:
          - packages/bootstrap
    steps:
      - uses: actions/checkout@v4
        name: Checkout repository
        with:
          fetch-depth: 0

      - uses: actions/setup-node@v4
        name: Setup Node.js @v20
        with:
          cache: npm
          node-version: 20.x

      - run: npm ci
        name: Install dependencies
        env:
          HUSKY: 0

        # Uses the official Cypress GitHub action https://github.com/cypress-io/github-action
      - uses: cypress-io/github-action@v6
        name: Cypress run
        with:
          # Starts web server for E2E tests - replace with your own server invocation
          # https://docs.cypress.io/guides/continuous-integration/introduction#Boot-your-server
          start: npm run dev
          wait-on: "http://localhost:3000" # Waits for above
          # Records to Cypress Cloud
          # https://docs.cypress.io/guides/cloud/projects#Set-up-a-project-to-record
          record: true
          parallel: true # Runs test in parallel using settings above
          working-directory: ${{ matrix.package }}
        env:
          # For recording and parallelization to work you must set your CYPRESS_RECORD_KEY
          # in GitHub repo → Settings → Secrets → Actions
          CYPRESS_RECORD_KEY: ${{ secrets.CYPRESS_RECORD_KEY }}
          # Creating a token https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          HUSKY: 0
