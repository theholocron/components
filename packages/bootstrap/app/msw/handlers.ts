import { handlers as authHandlers } from "../../page/inbox/auth/handlers";
import { handlers as taskHandlers } from "../../page/inbox/tasks/handlers";

export const handlers = [...taskHandlers, ...authHandlers];
