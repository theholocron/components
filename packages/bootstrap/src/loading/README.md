# Loading Component

A React Provider component for handling the loading of an application.

## Usage

```typescript
import { Loading } from "@theholocron/bootstrap";
import * as React from "react";

interface AppProps {
	Component: React.ReactNode;
	pageProps?: any;
}

export default function App(props: AppProps) {
	const { Component, pageProps } = props;

	return (
		<Loading.Provider>
			<Component {...pageProps} />
		</Loading.Provider>
	);
}
```

### `useLoading`

Use this function for getting the loading state and setting it.

```typescript
import { useLoading } from "@theholocron/bootstrap";

function Component() {
	const { isLoading, setLoading } = useLoading();

	const handleClick = () => {
		setLoading?.(true);
	};

	if (isLoading) {
		return <div>Loadingâ€¦</div>;
	}

	return (
		<button onClick={handleClick} type="button">
			Set Loading
		</button>
	);
}
```
