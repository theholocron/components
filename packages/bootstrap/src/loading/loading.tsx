import * as React from "react";
import type { WithChildren } from "../type";

export interface ILoading {
	isLoading?: boolean;
	setLoading?: (isLoading: boolean) => void;
}

const Context = React.createContext<ILoading | undefined>(undefined);
Context.displayName = "@theholocron/bootstrap/Loading";

export interface LoadingProps extends WithChildren {
	loader: React.ReactElement | React.ComponentType;
}

export function Provider(props: LoadingProps) {
	const { children, loader: Loader } = props;

	const [isLoading, setIsLoading] = React.useState<boolean>(false);

	const setLoading = React.useCallback((state: boolean) => {
		setIsLoading(state);
	}, []);

	const value = React.useMemo(() => ({ isLoading, setLoading }), [isLoading]);

	return (
		<Context.Provider value={value}>
			{isLoading ? <Loader /> : children}
		</Context.Provider>
	);
}
Provider.displayName = "@theholocron/bootstrap/Loading";

export function useLoading(): ILoading {
	const context = React.useContext(Context);

	if (context === undefined) {
		throw new Error("useLoading must be used within Loading.Provider!");
	}

	return context;
}
