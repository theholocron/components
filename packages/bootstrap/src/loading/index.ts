import { Provider } from "./loading.tsx";

export { useLoading } from "./loading.tsx";
export type { ILoading } from "./loading.tsx";

export default { Provider };
