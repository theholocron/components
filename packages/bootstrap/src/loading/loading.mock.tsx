import * as React from "react";
import { useLoading } from "./loading";

const spinAnimation = `
	@keyframes spin {
		0% { transform: rotate(0deg) translate3d(0, 0, 0); }
		100% { transform: rotate(360deg) translate3d(0, 0, 0); }
	}
`;

export function Loader () {
	return (
		<div
			data-testid="loader"
			style={{
				display: "inline-flex",
				alignItems: "center",
				justifyContent: "center",
				width: "100%",
				height: "100vh",
			}}
		>
			<style dangerouslySetInnerHTML={{ __html: spinAnimation }} />
			<svg
				fill="none"
				height={64}
				style={{
					animation: "spin 0.88s linear infinite",
					willChange: "transform",
				}}
				viewBox="0 0 48 48"
				width={64}
				xmlns="http://www.w3.org/2000/svg"
			>
				<path
					d="M48 24C48 37.2548 37.2548 48 24 48C10.7452 48 0 37.2548 0 24C0 10.7452 10.7452 0 24 0C37.2548 0 48 10.7452 48 24ZM4.8 24C4.8 34.6039 13.3961 43.2 24 43.2C34.6039 43.2 43.2 34.6039 43.2 24C43.2 13.3961 34.6039 4.8 24 4.8C13.3961 4.8 4.8 13.3961 4.8 24Z"
					fill="#2163e2"
					fillOpacity={0.15}
				/>
				<path
					d="M24 45.6C24 46.9255 25.0772 48.0125 26.3961 47.8802C31.8813 47.33 37.038 44.9031 40.9706 40.9706C45.4714 36.4697 48 30.3652 48 24C48 17.6348 45.4714 11.5303 40.9706 7.02944C37.038 3.09691 31.8813 0.67003 26.3961 0.119834C25.0772 -0.0124555 24 1.07452 24 2.4C24 3.72548 25.0787 4.78451 26.3938 4.94974C30.6034 5.47862 34.5447 7.39177 37.5765 10.4236C41.1772 14.0243 43.2 18.9078 43.2 24C43.2 29.0922 41.1771 33.9758 37.5765 37.5765C34.5447 40.6082 30.6034 42.5214 26.3938 43.0503C25.0787 43.2155 24 44.2745 24 45.6Z"
					fill="#2163e2"
				/>
			</svg>
		</div>
	);
}

// Example component that utilizes the Loading context
// @TODO: update to use msw
export const TestComponent = () => {
    const { setLoading } = useLoading();

	const fetchData = async () => {
		setLoading(true);
		await fetch("/api/data"); // Trigger a network request
		setLoading(false);
	};

    return (
        <div>
            <button onClick={fetchData}>Fetch Data</button>
            <p>Click the button to simulate loading.</p>
        </div>
    );
};

