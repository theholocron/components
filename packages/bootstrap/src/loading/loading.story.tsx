import type { Meta, StoryObj } from "@storybook/react";
import { expect, userEvent, within } from "@storybook/test";
import { delay, http, HttpResponse } from "msw";
import * as MockLoading from "./loading.mock";
import Loading, { useLoading } from "./index";

const meta = {
	argTypes: {
		loader: {
			description: "Loader component",
		},
	},
	component: Loading.Provider,
	title: "Loading",
} satisfies Meta<typeof Loading.Provider>;
export default meta;
type Story = StoryObj<typeof meta>;

export const Default = {
	args: {
		loader: MockLoading.Loader,
	},
	parameters: {
		msw: {
			handlers: [
				http.get("/api/data", async () => {
					await delay(2000);
					return HttpResponse.json({ message: "Success" });
				}),
			],
		},
	},
	render: (args) => (
		<Loading.Provider {...args}>
			<MockLoading.TestComponent />
		</Loading.Provider>
	),
	play: async ({ canvasElement }) => {
		const canvas = within(canvasElement);
		const fetchButton = await canvas.getByText("Fetch Data");

		// Simulate clicking the button
		await userEvent.click(fetchButton);

		// Expect the loader to be displayed
		const loader = await canvas.findByTestId("loader"); // Assuming your Loader component has this test ID
		expect(loader).toBeInTheDocument();

		// Wait for 2 seconds to simulate loading time
		await new Promise((resolve) => setTimeout(resolve, 3000));

		// Check that the loader is removed after loading
		expect(loader).not.toBeInTheDocument();
	},
} satisfies Story;

// Attempt to use the hook without provider
export const Error = {
	render: () => {
		const Component = () => {
			try {
				useLoading();
			} catch (error: unknown) {
				return (
					<div data-testid="error-message">
						{(error as Error)?.message}
					</div>
				);
			}
			return null; // Render nothing if no error
		};

		return <Component />;
	},
	play: async ({ canvasElement }) => {
		const canvas = within(canvasElement);

		const errorMessage = await canvas.findByTestId("error-message");
		expect(errorMessage).toHaveTextContent(
			"useLoading must be used within Loading.Provider!"
		);
	},
} satisfies Story;
