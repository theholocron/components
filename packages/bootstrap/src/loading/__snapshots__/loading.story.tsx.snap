// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading Default play-test 1`] = `
<div>
  <button>
    Fetch Data
  </button>
  <p>
    Click the button to simulate loading.
  </p>
</div>
`;

exports[`Loading Error play-test 1`] = `
<div data-testid="error-message">
  useLoading must be used within Loading.Provider!
</div>
`;
